packages/@aws-cdk/aws-eks/lib/user-data.ts

```diff
@@ -3,7 +3,7 @@ import { Stack } from '@aws-cdk/core';
 import { BootstrapOptions, ICluster } from './cluster';
 
 // eslint-disable-next-line max-len
-export function renderAmazonLinuxUserData(clusterName: string, autoScalingGroup: autoscaling.AutoScalingGroup, options: BootstrapOptions = {}): string[] {
+export function renderAmazonLinuxUserData(cluster: ICluster, autoScalingGroup: autoscaling.AutoScalingGroup, options: BootstrapOptions = {}): string[] {
```

Comment:
I just realized that for imported clusters, `cluster.clusterCertificateAuthorityData` might be undefined and will throw an exception.

Since this is only relevant for managed clusters at the moment, lets change to:

```suggestion
export function renderAmazonLinuxUserData(cluster: Cluster, autoScalingGroup: autoscaling.AutoScalingGroup, options: BootstrapOptions = {}): string[] {
```

Reply:
I have changed to `Cluster`, as you suggested - but I will also note that `renderBottlerocketUserData` uses `cluster.clusterCertificateAuthorityData` from `ICluster`.

Future TODO: Ideally we would turn `clusterCertificateAuthorityData==undefined` into empty string (or not provide the bootstrap.sh arg in this case), and then bootstrap.sh will fall back to the old behaviour of querying at runtime - but it's currently an untyped exception :(

Reply:
> Since this is only relevant for managed clusters at the moment, lets change to: ICluster->Cluster

.. No, wait.  Doesn't this need to work for imported clusters too?  I think `renderAmazonLinuxUserData` is used in building the worker node userdata.  And we definitely want to support adding worker nodes to clusters than came from `Cluster.fromClusterAttributes`.
 
(At least, I need to rewrite all the unittest fixtures for this function to stop using an imported `ICluster` - which I'm fine to do, but it made me stop and think.)

Reply:
Ok, so I've changed this to catch the 'is not defined for this imported cluster' exception and silently skip the bootstrap.sh option in each case.  This will fallback to the existing behaviour (bootstrap.sh will discover these values by calling DescribeCluster at instance boot time).

The "ignore this exception" code is a moderately ugly string match - I tried replacing this with a typed exception, but extending `Error` seems to upset JSII.  Another option is to introduce new `ICluster` `tryClusterEndpoint()`, etc methods that return `undefined` rather than throw - that's a fairly large addition to a public class, but let me know if you'd prefer that version.

Reply:
@anguslees 

> .. No, wait. Doesn't this need to work for imported clusters too? I think renderAmazonLinuxUserData is used in building the worker node userdata. And we definitely want to support adding worker nodes to clusters than came from Cluster.fromClusterAttributes

So...thats conceptually true, but notice that `user-data.ts`is not exported from `index.ts`, so its essentially a private API. And as far as our code goes, we only call it from `connectAutoScalingGroupCapacity`:

https://github.com/aws/aws-cdk/blob/567d64d70f92adbba9ff9981184d88b46fb95652/packages/%40aws-cdk/aws-eks/lib/cluster.ts#L1264-L1269

Which is only available on `Cluster`, not `ICluster`. I good with supporting `ICluster` if we can, but we should have an additional test to make sure we don't change the error message on one side only. 

> but I will also note that renderBottlerocketUserData uses cluster.clusterCertificateAuthorityData from ICluster

Yeap you're right, it should have been `Cluster` as well. Though I think that in this case, `cluster.clusterCertificateAuthorityData` has to be defined for the user-data to work, so while the context of the error won't be accurate, its ok for it to throw.